### PR TITLE
docs: Extended features in 5G SA documentation

### PR DESCRIPTION
**Pull request by sreedharkumartn**
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

## Summary
Documentation for Extended features in 5G SA

## Test Plan
Screenshots of the Documentation
image
image
image
image
image
image
image
image
image
image
image
image

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+id: extended_5G_sa_features
+title: Extended 5G SA Features
+hide_title: true
+---
+
+# Extended Features of 5G SA FWA
+
+Magma 5G SA feature enables operators to provide enhanced Mobile Broadband services and opens up new opportunites for business.
```
[the remark this answers is not in the document]
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+## SUCI Extensions
+
+5G standards mitigates the vulnerabilities caused due to IMSI catchers by introducing SUCI. SUCI is a privacy-preserving identifier containing the concealed SUPI.
+
+In this release we support two different profiles of SUCI Extensions. This document gives much information about [SUCI extensions](suci_extensions.md).
+
+## Feature parity with 4G - Stateless feature
+
+The MME will store UE registrations/session management data in-memory. Any MME failure or restart will result in the loss of all data, including UE registration and Operator Account information, preventing UEs from accessing packet core services. As a result, UEs will need to reconnect and re-register.
```
[the remark this answers is not in the document]
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    SessionD-->>PipelineD: Set SMF Sessions
+```
+
+Here, Rule 1 is the default QoS policy and Rule 2 is the QoS policy we are applying in Session Modification.
+
+Same sequence diagram is applicable for Deletion of QoS policy.
+
+### Limitations
+
+1. Currently we are supporting serial operation only, which means addition of a QoS policy or deletion of a QoS policy in a Session modification request. A single request containing add and delete is currently not supported.
```
Comment (ganeshg87) [inline docs/readmes/lte/extended_5G_sa_features.md]: replace "we are" with respect to AGW

Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
+---
+id: extended_5G_sa_features
+title: Extended 5G SA Features
+hide_title: true
+---
+
+# Extended Features of 5G SA FWA
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md:7]: The title says `5G SA FWA`, but here only `5G SA` is named.
- Can we explain what `SA` and `FWA` stands for in this document please.
- Can we consistently use `SA FWA` or only `SA` please. Unless there is a difference, however this difference should than be made clear.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
+---
+id: extended_5G_sa_features
+title: Extended 5G SA Features
+hide_title: true
+---
+
+# Extended Features of 5G SA FWA
+
+Magma 5G SA feature enables operators to provide enhanced Mobile Broadband services and opens up new opportunities for business.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: This is a meaningless sentence which should be removed from technical documentation.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+id: extended_5G_sa_features
+title: Extended 5G SA Features
+hide_title: true
+---
+
+# Extended Features of 5G SA FWA
+
+Magma 5G SA feature enables operators to provide enhanced Mobile Broadband services and opens up new opportunities for business.
+
+This document gives an overview of the Extended features added on top of [integrated 5G SA](https://docs.magmacore.org/docs/lte/integrated_5g_sa) such as 5G QoS, IPv6 & Dual IPv4v6 support, Network initiated Session Modification, SUCI Extensions, Stateless feature and GTP extension header support.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: - You could add relative links to the paragraphs in your list of features.
- You might want to consider a list instead. Kind of a table of contents.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+The QoS flow in PDU session accept message.
+![QoS in PDU Session Accept message](assets/lte/QoS_in_pdu_session_accept.png?raw=true "QoS in PDU Session Accept message")
+
+The QoS flow in PDU session modification command message.
+![QoS in PDU Session Modification Command message](assets/lte/QoS_in_pdu_session_modification_command.png?raw=true "QoS in PDU Session Modification Command message")
+
+## IPv6 & Dual IPv4v6 support
+
+IPv6 enables address availability for vast device growth in new areas of technology, such as loT. Large-scale device growth and deployment of loT devices that require less power, lower latency, and higher bandwidth/throughput are made possible by 5G.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: This has nothing to do with Magma and should not be part of technical documentation.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+This document gives an overview of the Extended features added on top of [integrated 5G SA](https://docs.magmacore.org/docs/lte/integrated_5g_sa) such as 5G QoS, IPv6 & Dual IPv4v6 support, Network initiated Session Modification, SUCI Extensions, Stateless feature and GTP extension header support.
+
+## 5G QoS
+
+The Quality of service (QoS) is the mechanism that works on a network to control traffic and ensure the performance of critical applications with limited network capacity. By giving particular high-performance applications priority, it enables the businesses to modify their overall network traffic.
+
+The QoS flow is the lowest level granularity within the 5G system and is where policy and charging are enforced.
+
+With the current infrastructure, AGW supports the QoS in this release for both PDU session accept and PDU session modification command messages.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: Can we have the acronym `PDU` explained here, please. It is the first occurrence of this acronyms in this document.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+Getting IPv6 address allocated in PDU Session accept message.
+![Allocation of IPv6 address](assets/lte/IPv6.png?raw=true "Allocation of IPv6 address")
+
+Getting IPv4v6 address allocated in PDU Session accept message.
+![Allocation of IPv4v6 address](assets/lte/IPv4v6.png?raw=true "Allocation of IPv4v6 address")
+
+## Network initiated Session Modification
+
+AGW supports one default PDU session. Network Initiated PDU Session Modification is one of the way to create Dedicated QoS Flows, if a new application function require to access a specific QoS policy but the current QoS policy cannot support services, it might be necessary to dynamically add a new QoS policy to an existing PDU session. In this case, PCF will send the session modification request to sessiond.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: - `requires`
- The first sentence is rather long and hard to understand :spaghetti:. Can we split this up?
- First occurance of `PCF`. Can we explain it, please?

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+Once the QoS policy is installed and its no longer required then PCF will trigger the request to clear the existing QoS policy from sessiond using session modification.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: `its` :arrow_right: `it is`

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+```
+
+Here, Rule 1 is the default QoS policy and Rule 2 is the QoS policy we are applying in Session Modification.
+
+Same sequence diagram is applicable for Deletion of QoS policy.
+
+### Limitations
+
+1. Currently AGW is supporting serial operation only, which means addition of a QoS policy or deletion of a QoS policy in a Session modification request. A single request containing add and delete is currently not supported.
+2. Right now this feature is still to be integrated with PCF. The testing has been using the stub CLIs in sessiond.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: The second sentence sound weird. Maybe `The tests used the stub CLI in sessiond.`?
Is it really multiple CLIs?

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+5G standards mitigates the vulnerabilities caused due to IMSI catchers by introducing SUCI. SUCI is a privacy-preserving identifier containing the concealed SUPI.
+
+In this release AGW supports two different profiles of [SUCI Extensions](https://docs.magmacore.org/docs/next/lte/suci_extensions).
+
+## Stateless feature
+
+The MME will store UE registrations/session management data in-memory. Any MME failure or restart will result in the loss of all data, including UE registration and Operator Account information, preventing UEs from accessing packet core services. As a result, UEs will need to reconnect and re-register.
+
+Stateless feature provides failure recovery mechanism by storing user data in a persistent storage.
+On occurrence of a failure, AGW will retrieve the UE data from the persistent storage, a partial service interruption may occur. However, once the AGW has been respawned then registered users can access the service without having a fresh registration.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: - First sentence are actually two sentences in one. Split after `persistent storage`.
- `then` in second sentence should be replaced by `,`.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+The MME will sync UE registration and session management data into persistent storage (Redis Database) for pre-defined trigger points. Incase of any MME failure/restart, MME will fetch UE specific data from Redis Database.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: `In case`

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+![Stateless Feature](assets/lte/Stateless_feature.png?raw=true "Stateless Feature")
+
+## GTP extension header support
+
+The Extension Header Length field specifies the length of the particular Extension header in 4 octet units. The Next Extension Header Type field specifies the type of any Extension Header that may follow a particular Extension Header. If no such Header follows, then the value of the
+Next Extension Header Type shall be 0.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: - superflous line break.
- `shall be 0` sounds a bit ambiguous to me. What is meant? Should, must, can, will?

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+     | Ext-Hdr Length|                                               |
+     +-+-+-+-+-+-+-+-+                                               |
+     |                  Extension Header Content                     |
+     .                                                               .
+     .                                               +-+-+-+-+-+-+-+-+
+     |                                               |  Next-Ext-Hdr |
+     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
+
+In 5G SA, each flow is forwarded based on the appropriate QoS rules. QoS rules are configured by SMF as QoS profiles to UP components and these components perform QoS controls to PDUs based on rules. In downlink, a pipelineD pushes QFI into an extension header, and transmits the PDU to RAN.  In uplink, each UE obtains the QoS rule from SMF, and transmits PDUs with QFI containing the QoS rules to the RAN.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: There is a double whitespace here.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+session_manager_cli.py policy_rar IMSI001019976545660 IMSI001019976545660-112383 INTERNET ADD 10 6 "UL,6,192.168.128.12,0,192.168.200.1,12345;DL,1,8.8.8.8,0,192.168.128.12,0" "1000000,2000000,1000000,2000000,15,1,0"
+```
+
+Stub CLI used to delete the QoS policy in Network initiated session modification.
+
+```text
+session_manager_cli.py policy_rar IMSI001019976545660 IMSI001019976545660-112383 INTERNET DEL 10 6 "UL,6,192.168.128.12,0,192.168.200.1,12345;DL,1,8.8.8.8,0,192.168.128.12,0" "1000000,2000000,1000000,2000000,15,1,0"
+```
+
+Here is the details of the [Stub CLI command](assets/lte/Session_modification_stub_cli_commands "Stub CLI Command") for add or delete the QoS policy in Network initiated session modification.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: - `Here are the details`
- The link seems broken.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+### Stateless feature
+
+To support stateless feature first should enable config_stateless flag.
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: Subject missing in this sentence.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+magmad_cli.py config_stateless check_stateless
+```
+
+The above command should give `AGW Mode: STATELESS`.
+
+## Common Issues and Troubleshooting
+
+GTP tunnel not added in OVS table
+
+- Description :-
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: `:-` looks funny. What is that supposed to do?
Rather use `Description: Your description text here`.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+  GTP tunnel not added in OVS table due to some parameters missing in `/etc/magma/pipelined.yml` file.
+
+- Solution :-
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: `:-` looks funny. What is that supposed to do?

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+  The following parameters should add in `/etc/magma/pipelined.yml` file
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: This is an unclear and incomplete sentence.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+
+```text
+  #Internal port for processing internal sampling packets
```
Comment (Neudrino) [inline docs/readmes/lte/extended_5G_sa_features.md]: Inconsistent white-spacing in comments.

**Review by Neudrino — COMMENTED** [review]
Thanks for the documentation contribution.
There are some formal aspects which need to be corrected.
In general it would be good, if for any abbreviations used in the document the long form would be given on first occurance.

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 4 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md:7]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 5 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 7 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 8 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 9 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 10 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 11 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 12 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 13 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 14 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 15 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 17 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 18 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 19 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 20 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 21 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 6 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed Review Comment

Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: image

File: docs/readmes/lte/extended_5G_sa_features.md (hunk 16 above)
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: image

Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: On clicking the "Stub CLI Command", a document (`docs/readmes/assets/lte/Session_modification_stub_cli_commands`) will be downloaded.

Comment (Neudrino) [reply docs/readmes/lte/extended_5G_sa_features.md]: You are right. The "tooltip function" for markdown links is not mentioned within the [docusaurus documentation](https://docusaurus.io/docs/markdown-features/assets), such that I was unaware it is indeed supported.

**Review by Neudrino — DISMISSED** [review]
Thanks a lot for all the corrections! :+1: :arrow_right: Good to go for me.

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+Here, Rule 1 is the default QoS policy and Rule 2 is the QoS policy configured as part of Session Modification. Same sequence diagram is applicable for Deletion of QoS policy.
+
+### Limitations
+
+1. Currently only CLI based Session modification configuration is supported.
+2. As part of the current release, feature is not tested end to end with PCF (Policy Control Function).
+
+## SUCI Extensions
+
+In compilance with 5G standards and to provide enhanced security, SUCI based registration has been introduced in the current release supporting both Profile-A and Profile-B configuration using ECIES-based protection scheme. With this feature end devices (UE/CPE) can encrypt the identifiers (IMSI) during the registration request using the pre-shared keys which will be then decrypted and processed by AMF.
```
[the remark this answers is not in the document]
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed

File: docs/readmes/lte/extended_5G_sa_features.md [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+
+The above command should give `AGW Mode: STATELESS`.
+
+## Common Issues and Troubleshooting
+
+- Traffic related Issues in [common path](https://github.com/magma/magma/blob/master/docs/readmes/howtos/troubleshooting/datapath_connectivity.md).
+
+- Ping traffic not working: with QFI support in GTP-U extension header this parameter needs to be present in all traffic cases otherwise the default rule is to drop the packet.
+
+- IPv6 feature is supported only on Non-NAT mode.
```
[the remark this answers is not in the document]
Comment (sreedharkumartn) [reply docs/readmes/lte/extended_5G_sa_features.md]: Addressed

**Review by Neudrino — APPROVED** [review]

**Review by panyogesh — APPROVED** [review]
LGTM

**Pull request merged**